### PR TITLE
Add move constructors for `string_scalar` and `struct_scalar`

### LINKED ISSUE #8427: [FEA] Construct scalar by moving a free buffer
Currently, scalar objects are constructed by some ways such as copying from another scalar object, or constructed by copying data from a representative object (for example, a string scalar can be constructed from a string_view). However, they (scalars) cannot be constructed from an orphaned buffer. Sometimes, we may want to create a scalar from some computation, which results in a large buffer (especially `string_scalar`), but we can only construct a scalar by copying that buffer into the scalar internal buffer. This is a waste of resources.

Solution: Add constructors for some scalar classes (such as `string_scalar`) that accept a free buffer and move it into the internal buffer.

### PR DESCRIPTION
**Pull request by ttnghia**
This PR adds some move constructors for `string_scalar` and `struct_scalar` that accept r-value references to existing buffers. By doing so, the input buffer is moved into the internal buffer of the newly constructed scalars instead of copying.

This also cleans up/rewrites doxygens for `scalar.hpp`.

Closes #8427.

### CODE REVIEW

File: cpp/include/cudf/scalar/scalar.hpp [hunk 1]
```diff
@@ -395,81 +395,95 @@ class string_scalar : public scalar {
...
    */
   string_scalar(rmm::device_scalar<value_type>& data,
                 bool is_valid                       = true,
                 rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
                 rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
   /**
-   * @brief Implicit conversion operator to get the value of the scalar in a host std::string
+   * @brief Construct a new string scalar object from an orphaned buffer in device memory.
+   * Note that this function move the orphaned buffer into the internal buffer.
```
Comment (davidwendt) [inline cpp/include/cudf/scalar/scalar.hpp:449]:
```suggestion
   * 
   * Note that this constructor moves the orphaned buffer into the internal buffer.
```

File: cpp/include/cudf/scalar/scalar.hpp [hunk 2]
```diff
@@ -689,37 +703,51 @@ class struct_scalar : public scalar {
...
    */
   struct_scalar(host_span<column_view const> data,
                 bool is_valid                       = true,
                 rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
                 rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
   /**
-   * @brief Returns a non-owning, immutable view to underlying device data
+   * @brief Construct a new struct scalar object from an orphaned table in device memory.
+   * Note that this function move the orphaned table into the internal data table.
```
Comment (davidwendt) [inline cpp/include/cudf/scalar/scalar.hpp:737]:
```suggestion
   * 
   * Note that this function moves the orphaned table into the internal table member.
```

**Review by davidwendt — CHANGES_REQUESTED** [review]
Why are all the comments reformatted?

**Comment by ttnghia** [comment]
> Why are all the comments reformatted?

I just want to enforce aesthetic consistency for them: some comments ended with a period (`.`) while some others don't.

File: cpp/include/cudf/scalar/scalar.hpp [hunk 3]
```diff
@@ -56,43 +55,44 @@ class scalar {
    *
    * @param[in] other The scalar to copy.
    * @param[in] stream CUDA stream used for device memory operations.
-   * @param[in] mr Device memory resource to use for device memory allocation
+   * @param[in] mr Device memory resource to use for device memory allocation.
```
Comment (harrism) [inline cpp/include/cudf/scalar/scalar.hpp:58]: While you are adding periods, you could remove the [in]s.

File: cpp/include/cudf/scalar/scalar.hpp [hunk 4]
```diff
@@ -395,81 +395,96 @@ class string_scalar : public scalar {
...
    */
   string_scalar(rmm::device_scalar<value_type>& data,
                 bool is_valid                       = true,
                 rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
                 rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
   /**
-   * @brief Implicit conversion operator to get the value of the scalar in a host std::string
+   * @brief Construct a new string scalar object from an orphaned buffer in device memory.
```
Comment (harrism) [inline cpp/include/cudf/scalar/scalar.hpp:448]: Let's not add potentially confusing terminology ("orphaned"). 
```suggestion
   * @brief Construct a new string scalar object by moving an existing string data buffer.
```

File: cpp/include/cudf/scalar/scalar.hpp [hunk 5]
```diff
@@ -395,81 +395,96 @@ class string_scalar : public scalar {
...
+   *
+   * Note that this constructor moves the orphaned buffer into the internal data buffer.
```
Comment (harrism) [inline cpp/include/cudf/scalar/scalar.hpp:450]:
```suggestion
   * Note that this constructor moves the existing buffer into the internal data buffer; no copy is performed.
```

File: cpp/include/cudf/scalar/scalar.hpp [hunk 6]
```diff
@@ -689,37 +704,52 @@ class struct_scalar : public scalar {
...
+   * @param mr Device memory resource to use for device memory allocation.
    */
   struct_scalar(host_span<column_view const> data,
                 bool is_valid                       = true,
                 rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
                 rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
   /**
-   * @brief Returns a non-owning, immutable view to underlying device data
+   * @brief Construct a new struct scalar object from an orphaned table in device memory.
```
Comment (harrism) [inline cpp/include/cudf/scalar/scalar.hpp:737]:
```suggestion
   * @brief Construct a new struct scalar object from an existing table in device memory.
```

File: cpp/include/cudf/scalar/scalar.hpp [hunk 7]
```diff
@@ -689,37 +704,52 @@ class struct_scalar : public scalar {
...
+   *
+   * Note that this constructor moves the orphaned table into the internal data table.
```
Comment (harrism) [inline cpp/include/cudf/scalar/scalar.hpp:739]:
```suggestion
   * Note that this constructor moves the existing table data into the internal data table; no copies are performed.
```

**Review by harrism — CHANGES_REQUESTED** [review]
I think secretly this PR is more about adding periods to the comments. :)

I think this is a good addition, just some doc suggestions.

**Comment by harrism** [comment]
> Other (line adjustment) may be due to some explicit reformatting done by my IDE.

Please disable this and just use our mandated clang-format settings. Other than automatic line-wrapping to ensure passing clang-format tests, any reformatting of comments should be done manually and not on every file save or it will make PRs harder to review.

I had a hard time getting to the real changes in this PR.

**Comment by ttnghia** [comment]
> Please disable this and just use our mandated clang-format settings. Other than automatic line-wrapping to ensure passing clang-format tests, any reformatting of comments should be done manually and not on every file save or it will make PRs harder to review (and cause more conflicts, e.g. with my in-progress scalar.hpp refactor PR)
> 
> I had a hard time getting to the real changes in this PR.

I'm sorry about that. Typically it is disabled, I just accidentally called that this time.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
LGTM.

(Sorry, I thought I had already approved this PR last week.)

**Review by davidwendt — APPROVED** [review]

**Review by harrism — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**